finance_dl/discover.py

```diff
@@ -0,0 +1,176 @@
+"""Retrieves transaction and balance information from Discover.
+Logs into the web interface, downloads an OFX file.
+
+This uses the `selenium` Python package in conjunction with `chromedriver` to
+scrape the Discover website.
+
+Configuration:
+==============
+
+The following keys may be specified as part of the configuration dict:
+
+- `credentials`: Required.  Must be a `dict` with `'username'` and `'password'`
+  keys.
+
+- `output_directory`: Required.  Must be a `str` that specifies the path on the
+  local filesystem where the output will be written.  If the directory does not
+  exist, it will be created.
+
+- `profile_dir`: Optional.  If specified, must be a `str` that specifies the
+  path to a persistent Chrome browser profile to use.  This should be a path
+  used solely for this single configuration; it should not refer to your normal
+  browser profile.  If not specified, a fresh temporary profile will be used
+  each time.  It is highly recommended to specify a `profile_dir` to avoid
+  having to manually enter a multi-factor authentication code each time.
+
+Output format:
+==============
+
+OFX file.
+
+Example:
+========
+
+    def CONFIG_discover():
+        return dict(
+            module='finance_dl.discover',
+            credentials={
+                'username': 'XXXXXX',
+                'password': 'XXXXXX',
+            },
+            output_directory=os.path.join(data_dir, 'discover'),
+            # firefox seems to work better on Discover's website
+            profile_dir=os.path.join(profile_dir, 'firefox'),
+        )
+
+Interactive shell:
+==================
+
+From the interactive shell, type: `self.run()` to start the scraper.
+
+"""
+
+import urllib.parse
+import re
+import datetime
+import time
+import logging
+import os
+import shutil
+from selenium.common.exceptions import NoSuchElementException, TimeoutException
+from selenium.webdriver.common.keys import Keys
+
+from . import scrape_lib
+
+
+logger = logging.getLogger('discover_scrape')
+
+netloc_re = r'^([^\.@]+\.)*discover.com$'
+XPATH_OF_LAST_FOUR_DIGITS = "/html/body/div[1]/main/div[3]/div[2]/div/div[1]/a/span[1]"
```

Comment:
You might find it more reliable to use a different way of finding the account number --- this xpath expression seems like it may be invalidated by even minor changes to the page.

Reply:
In fact maybe you can read the account number from the downloaded OFX file instead?

Reply:
Changed. Hopefully more robust.